### ISSUE TITLE
[Forgejo Migration] Migrate GitHub issues to Forgejo

### ISSUE THREAD

**Issue by foss-** [friendica/friendica#13391]
Part of Forgejo Migration: Migrate GitHub issues to Forgejo

Depends on https://codeberg.org/forgejo/forgejo/issues/340
Blocked by https://forum.forgefriends.org/t/about-the-friendly-forge-format-f3/681 or https://github.com/go-gitea/gitea/pull/18165
Resolves https://github.com/friendica/friendica/issues/9834

**Comment by foss-** [comment]
The good news is, @OrganicMaps has just fully moved to Forgejo as per https://mastodon.social/@organicmaps/114233788700982882

- 54k commits
- 9.5k issues
- 4.3k pull requests
- 100k comments

Meaning whatever got stuck during the 2023 attempt to move Friendica to codeberg may no longer be an issue or they just used another database format 🤔. Call for volunteers to give the migration another go I guess 🫣

**Comment by annando** [comment]
Before we do the move, I would like to check if its possible to have got some server structure there with at least two servers with some fallback/synchronization mechanism, so that the repository will work, even when one of the servers fail. it will happen from time to time that the server isn't working totally fine. And since we are all doing it in out spare time, no one can expect that problems are fixed in minutes or even hours. So it would be great to have some fallback mechanism.

BTW: I really don't know if Forgejo can do something like that and also I'm not admin enough to even know if this is possible at all.

**Comment by tobiasd** [comment]
I'll do a test migration later today.

**Comment by MrPetovan** [comment]
@annando We already have a GitHub mirror of the addons repository, we can also keep the core GitHub repository as a read-only mirror of the ForgeJo repo.

**Comment by Art4** [comment]
> Before we do the move, I would like to check if its possible to have got some server structure there with at least two servers with some fallback/synchronization mechanism, so that the repository will work, even when one of the servers fail. it will happen from time to time that the server isn't working totally fine. And since we are all doing it in out spare time, no one can expect that problems are fixed in minutes or even hours. So it would be great to have some fallback mechanism.

Very good point. The reliability is very important and as we can see with the addon repository sometimes the server no longer responds. What do you think about switching to codeberg instead of own servers?

**Comment by MrPetovan** [comment]
I don't think we should switch from a third-party platform to another third-party platform, but I wouldn't mind having several read-only mirrors on popular platforms like GitHub or Codeberg.

**Comment by marcusxss** [comment]
Very happy to see an issue for this!

If needed I'd gladly try to help out with the migration. Hopefully it would inspire other open source projects to do the same, such as weblate, who I notice have linked to this issue.

I'm already running my own forgejo instance, though it's just for personal use so I haven't had much need to really dive into its features yet. Codeberg I've yet to try out but I'd use that as a mirror personally, or secondarily GitLab. GitLab uses a different CI syntax than GitHub though.

**Comment by annando** [comment]
My big problem with the move is stability. We all have got our daytime jobs and we have got other stuff that we do in our spare time as well. We see it with the pipeline and the addon repository that it crashes from time to time. And then it can take some hours until the pages run again. It is impossible for us to guarantee a 24/7 availability, unless we would be able to pay someone for this - which we can't.

The only solution that I see is some kind of load balancer combined with several git instances (Ideally run by different people) that are synchronised. And I really don't know how how to do this. This is way out of my abilities.

**Comment by marcusxss** [comment]
## Self hosted

It sounds like either the server isn't powerful enough or it's an issue of configuration?? Or maybe the software is unstable and it needs to have some sort of "self healing" set up? If it self healing, that sounds a bit like Kubernetes to me, but I'm not that familiar with Kubernetes personally. I've used Docker and Podman a little - but not yet Kubernetes. Podman can start kubernetes clusters, though. But kubernetes would be more taxing for the server, and not knowing the server I don't know if it can handle it?

Alternatively I'm wondering if it could be possible to have the pipeline/runners somewhere else, and then only have the repo, issues, pull requests and such to deal with on this server.

As for paying for server costs, if there was something similar to Patreon (but something that takes a smaller cut), perhaps some of the thousands of users would be willing to help pay for hosting the code. I remember someone asked about donating to Friendica not that long ago, so at least some people seem interested in it.

## Codeberg

Otherwise perhaps Codeberg is the way to go after all, as @Art4 suggested.
IMO anything is better than GitHub, and it would be unfortunate if the lack of a perfect solution should stand in the way of something that isn't perfect but still better.

And Codeberg is a good solution IMO. It would be hosted in Europe, and specifically Germany, where the most devs are located, where friendi.ca is also hosted. It would also make everyone much more familiar with Forgejo since it's mostly Forgejo under the hood, in case we wanted a switch to Forgejo in the future, and then Codeberg could become the mirror.

Edit: Oh, and I would also **gladly** help in migrating to Codeberg as well, in case that's wanted.

**Comment by marcusxss** [comment]
Turns out there are various third party runners for GitHub flavoured CI:
https://github.com/neysofu/awesome-github-actions-runners
...but it would come with a cost, perhaps because Microsoft wants the code on GitHub so badly, so they can train their LLM on it that they're willing to lose some money on it. And also, in general, so they can maintain control of the open source community.
Alternatively maybe the pipeline could run on Codeberg even if the code lives on our own Forgejo.

The reason I was thinking it might be relevant to run the CI elsewhere is since I imagine it's the most resource hungry component, but then I remembered that the CI is already running on friendi.ca. I'm unsure if it's the full CI, and GitHub runs the same, or if it's only part of the CI running on friendi.ca?

# More on importing issues (and other data)

I think either Codeberg or Forgejo should be able to import the issues with a GitHub Access Token. I've created an issue to Forgejo for more detailed documentation, as I'm unsure of whether any token can do so for a public repo, or if it needs to be a token for a user/organisation with some relation to the repo, like being a member, owner of it or something. I assume a read only token is enough. UPDATE: See edit below.

If it's a `fine-grained` token rather than a `classic` one, from looking at GitHubs documentation I *think* these are the needed permissions:

| Migrating what | Read access required for |
|---|---|
| Issues | Issues |
| Pull requests | Pull requests |
| Milestones | Issues |
| Labels | Issues |
| Releases | Contents |

Edit: It turns out that for a public repo and issues, at least, the personal access token can be from an account that's not a member of the repo/organization.

**Comment by marcusxss** [comment]
A (mostly) inspiring case: 
Zig just moved to Codeberg (Forgejo-based).
- https://ziglang.org/news/migrating-from-github-to-codeberg/
- https://www.theregister.com/2025/12/02/zig_quits_github_microsoft_ai_obsession/
- https://codeberg.org/ziglang/zig (Codeberg is currently experiencing a DDOS, though)

...but they decided not to migrate their 30000 issues (about 3000 open ones) currently:
> In the end, we opted for a simple strategy, sidestepping GitHub’s aggressive vendor lock-in: leave the existing issues open and unmigrated, but start counting issues at 30000 on Codeberg so that all issue numbers remain unambiguous. Let us please consider the GitHub issues that remain open as metaphorically “copy-on-write”. Please leave all your existing GitHub issues and pull requests alone. No need to move your stuff over to Codeberg unless you need to make edits, additional comments, or rebase. We’re still going to look at the already open pull requests and issues; don’t worry.

Not sure how issues migration was blocked specifically, though. Maybe they have too many, or comments or something aren't carried over? Not sure if one can only migrate open issues somehow?
But their approach would be another way to do so:
Leave old issues and PRs here, maybe stop the creation of new ones here (if possible), instead referring to creating them on the new one. Then at some point later it could be archived here as well?

That said, git.friendi.ca is currently returning "bad gateway" when nginx tries to connect to, I assume, forgejo? But I imagine it's fixable and that it could be reduced in frequency and I would gladly help investigate it.  I wonder if the server just isn't powerful enough for the peak demands on it. Or is it forgejo or specifically the CI part that's unstable?

**Comment by annando** [comment]
@mfxa git.friendi.ca runs on an own machine using Gitea.

**Comment by marcusxss** [comment]
Is there anything I can do to help with/accelerate this migration?
I've been planning to make this my highest priority after the recent release, but hadn't gotten around to following up.
Also I'll help if the target is Codeberg or Source Hut or something instead as well. I would just like to stop supporting closed source Microsoft GitHub.

**Comment by tobiasd** [comment]
Upstream (forgejo) has closed the blocking issue mentioned in the 1st post of this thread and it is still not possible to move the project including the issues to our own forgejo.

**Comment by foss-** [comment]
> Upstream (forgejo) has closed the blocking issue mentioned in the 1st post of this thread and it is still not possible to move the project including the issues to our own forgejo.

Not possible how or why? OrganicMaps was temporarily moved to Codeberg or a Forgejo instance and that did work fine with thousands of issues, which iirc was the blocking part back when moving (or copying friendica) there was first attempted a long while ago.

**Comment by tobiasd** [comment]
I tried moving the repository the other week (and several times, also to different forgejo / gitea instances for testing), it ends with error messages.

**Comment by marcusxss** [comment]
A bit disappointing that they just closed the issue.

...but very glad to hear someone's been prioritizing this, thanks!

Was it the same error message mentioned in the issue in the 1st post?:
```
Migrate repository from https://github.com/friendica/friendica failed: comment references non existent IssueIndex 22
```

Out of curiosity, I wonder if imports to Codeberg would work, for testing, unless that's one of the options you already tried. Being forgejo-based it likely won't, but maybe this is an area where they're different.

~~If someone could give me a **read only** access token, either a classic one or one with the permissions listed here (or at least "Issues")~~ (turns out that's not needed):
https://github.com/friendica/friendica/issues/13391#issuecomment-3228239103
...I can *also* try to migrate them to my own forgejo instance or codeberg and try to get in touch with the devs about it when it fails, if I'm not able to figure something out? Maybe specifically forgejo-next (the version in development?), as I see they suggested when closing it, so they can be sure it's unsolved.

Maybe I'd run into the rate limiting, though. I see @nupplaphil tried from a "premium" account previously to get around this, but still ran into an issue at that time. I don't have that. Hopefully one can specify a maximum number of requests in a given period of time.

Alternatively I may have found a tool that is able to fetch the issues to CSV:
https://github.com/gavinr/github-csv-tools
...but then the question is if there's a good way to import them, that's hopefully easier than directly inserting them into Forgejo's database.

**Comment by marcusxss** [comment]
I just noticed that IF we could successfully migrate them to Codeberg - if their migration process is slightly different/improved,  it's the importing server running out of resources (timeout or similar) or their support is perhaps better able to help - Forgejo has the feature to import them from Codeberg.

Then afterwards perhaps Codeberg could be the new mirror instead of GitHub.

**Comment by marcusxss** [comment]
It seems there's been some work recently on GitHub imports:
This was for v15 but backported to v14: https://codeberg.org/forgejo/forgejo/pulls/10798
This was for v15 but backported to v14: https://codeberg.org/forgejo/forgejo/pulls/10846
This was for v14, and seemingly not backported: https://codeberg.org/forgejo/forgejo/pulls/9348

v14 is the latest release. git.friendi.ca runs v13. Maybe it would be worth a try to update it to the latest v14, to get all of those fixes?

**Comment by tobiasd** [comment]
codeberg still fails

```
Error 1213 (40001): Deadlock found when trying to get lock; try restarting transaction
```

**Comment by marcusxss** [comment]
Ah alright. I wonder if the folks at Codeberg can provide more info on what's going wrong?

During some testing today it seems i've managed to migrate 4137 issues (at least there's that number of rows in the `issue` table in the database) to my own instance that's running V14, and it seems there's 5191 total issues when it's open + closed. (do we want closed as well or just open?)

Then I ran into the rate limit of 5000 requests per user per hour, which I just learned about:
https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api
I think it must be fetching the comments separately, or something, and that's why it's over 5000 at 4137 issues?

I will try to retry when the limit resets.
I hope the retry is able to pick up where it got to. Otherwise I'll try to figure out where to put in delays between requests so I don't exceed that limit.

<img width="649" height="178" alt="Image" src="https://github.com/user-attachments/assets/2747221b-67ab-48dc-aa7a-cd76e4e8614f" />

**Edit:**
Okay, retrying does not seem to work, as no further issues were imported, so maybe it IS trying to start over?:
<img width="649" height="178" alt="Image" src="https://github.com/user-attachments/assets/5085a325-e6fd-4794-b960-69b0945b44d2" />

Future note: Retrying starts fully over, so if something was written to the database you'll get integrity errors related to uniqueness trying to write values to the database that already exist.

Trying to hear if someone in their Matrix has suggestions, or otherwise I will try traffic shaping manually putting delays in between the requests. Then maybe it will take some hours but that's fine.
If successful git.friendi.ca could import them from my instance (edit 2: or upgrade to  V14.0.2 and try again)

**Edit 2:**
Exciting news. I was running 14.0.1, but I was told by one of their devs that one of the issues I listed above - a fix for the timeouts - was added in 14.0.2!
So just upgraded forgejo to that version and now I'm trying again.

**Comment by marcusxss** [comment]
I've attempted to migrate maybe around 10 times but so far they've failed.
The Forgejo team has fixed the issue with rate limiting <3, though, and they told me the unique constraint error is a result of choosing "retry" when data has already been written to the database. Currently retrying doesn't work, so the solution is to delete the repo and start anew.

I've created a new issue with the Forgejo team here:
https://codeberg.org/forgejo/forgejo/issues/11556

I will continue to test on my own. Fx. I'm noticing it tries to import:

- Open issues + closed issues
- Comments on issues
- Reacts on issues and comments

...but maybe all of those aren't that important? Fx. what if we only imported open issues and their data?
Or what if we skipped reactions?

I also think that if we could migrate open issues and closed issues separately, the likelihood of success would probably go up when less data is migrated in one go. The GitHub API supports getting open issues only, so maybe the library they use to import - https://github.com/google/go-github/ - also supports migrating only open/closed issues separately.

**Comment by tobiasd** [comment]
> ...but maybe all of those aren't that important? Fx. what if we only imported open issues and their data?
Or what if we skipped reactions?

I think open and closed issues are important as references to the work that has been done and the discussion around them as the github repository might just vanish one day.

**Comment by MrPetovan** [comment]
Reactions can definitely be skipped but comments may be important.

**Comment by marcusxss** [comment]
Some updates:

I've succesfully migrated the entire repo - including issues - to GitLab!:
https://gitlab.com/test8474611/friendica/-/work_items
Now Forgejo also has an option to import from GitLab, so I'm hoping imports from there work better.

I was looking into the Forgejo code and saw some internal options to "SkipReactions", and also adjust the page size, which should help with the most recent error I kept running into:
> error while listing repos: GET https://api.github.com/repos/friendica/friendica/issues/comments?direction=asc&page=1112&per_page=27&sort=created: 422 In order to keep the API fast for everyone, pagination is limited for this resource. []

I *think* I successfully modified the code so ~~comments~~ reactions were skipped and I lowered the page size which should slow down the migration but decrease the risk of running into the error above, but at 50 items per page, at least, it still failed. That said I haven't found a good way to verify that my changes took effect, because the resulting build is a binary.

...but when searching for that error I found some posts on import issues from GitLab, and read their docs a bit. They mention they successfully migrated "kubernetes" at:
> 80,000 pull requests.
> 45,000 issues.
> Approximately 1.5 million comments.

...which took 76 hours, so I thought it should be able to handle Friendica.
And it was.

**Comment by marcusxss** [comment]
Now trying to import from GitLab to git.friendi.ca:
<img width="661" height="471" alt="Image" src="https://github.com/user-attachments/assets/e88ed99f-ce7c-4885-9d7a-14dbdf4a19e3" />

We'll see.

**Comment by marcusxss** [comment]
Issues have now been migrated, at least as a test, to git.friendi.ca!
https://git.friendi.ca/friendica-test2/friendica/issues

Edit: For some reason it's saying all issues have the same author, even though in gitea that wasn't the case?!

foss pointed out that because of the gitlab import the usernames are pretty messed up, which is something that happens when gitlab imports them. It seems one can fix it somewhat, by mapping the imported user to a gitlab user, but that only works if a given user is on gitlab...
Not sure how it would look if direct import from GitHub to Forgejo actually succeeded.
I'm not sure if we can live with that username issue? Also not sure if we should use my import or do a new import into the existing friendica mirror repo, or something under the existing friendica organisation on git.friendi.ca? Or maybe we can move it?

**Experiments leading up to it:**

1. Import from GitLab to git.friendi.ca failed (only tried it once, though: "runtime error: index out of range")
2. Import from Gitlab to my own forgejo failed (only tried it once, though: I think it was the same error as above)
3. Spun up a gitea instance for testing
4. Import from GitHub to Gitea failed (github rate limiting)
5. Import from GitLab to Gitea failed in the first attempt (unique constraint error)
   **but the second attempt, this time only of issues, succeeded!**

Finally I then tried importing from gitea to forgejo, and it worked!

So the import process was:
GitHub -> GitLab -> Gitea -> Forgejo
...quite convoluted. ;) Hopefully Forgejo's GitHub import will be able to do the import directly in the future, at least for other projects.

Edit: Also labels weren't migrated from gitlab to gitea, so they're missing from the issues.
When I tried to migrate everything from gitlab to gitea it failed, but I could try to include labels.
Trying that now.

Edit 2: Migration with labels to gitea succeeded! Migrating from there to git.friendi.ca successful as well. Updated the repo link at the top.

**Comment by marcusxss** [comment]
Now I managed to do a migration of issues and labels directly from GitHub to Forgejo, so usernames aren't messed up!
It was to my own instance with a binary that seems to've been successfully built with modifications to skip reactions and to use a much lower page size, based on forgejo version 14.0.3. I can send the binary and you can replace it, or I if I get temporary server access I can replace it myself.

In case it's simply working because of some change in github instead of forgejo I just started another direct test migration to git.friendi.ca as is.

**Comment by marcusxss** [comment]
The direct migration from github to git.friendi.ca still failed.
I also tried to do a migration from my instance - with all the usernames as they should be - to git.friendi.ca, but after the migration all issues have "Ghost" as the author for some reason:

<img width="356" height="52" alt="Image" src="https://github.com/user-attachments/assets/88854bef-13e4-47e6-ba47-83491da6832a" />

It's here: https://git.friendi.ca/friendica-test/friendica/issues

So it seems we need to do the migration directly from github to git.friendi.ca to preserve usernames, and we should be able to with the same version and same changes I applied.
Let me know how you want to do/get/create the patched version mentioned before.

**Comment by MrPetovan** [comment]
What binaries need to be updated?

**Comment by marcusxss** [comment]
Argh, I just noticed comments are missing in the most recent imports: 
http://forgej.ydns.eu/friendica/friendica2/issues
but they ARE present in the github->gitlab import, and they are also there in the follow-up import from gitlab to gitea:
http://forgej.ydns.eu:3009/friendica2/friendica/issues
Now I don't remember if they were carried over from gitea to forgejo. I've since deleted that testing repo to clean up from all the attempts..

On the other hand, usernames are as they should be migrating directly from github to forgejo (1st link), but once they go through gitlab (2nd link) and especially go from forgejo -> forgejo they're not.

~~I'm wondering if maybe when reactions are skipped, comments are skipped as well, but now I'm not sure.~~
~~I/We could try a migration without skipping reactions and just lowering page size.~~ (wrong theory, see edit below)

--- 

> What binaries need to be updated?

Just one. It seems basically all forgejo's backend builds to a single file, and if using the docker version it's inside the container at:
/app/gitea/gitea
...then there's a symlink to it called forgejo:
```sh
$ which forgejo
/usr/local/bin/forgejo
$ ls -l /usr/local/bin/forgejo
lrwxrwxrwx    1 root     root             5 Jan 29 07:41 /usr/local/bin/forgejo -> gitea
```
...and `gitea` is a small script that calls `/app/gitea/gitea`.

The gitea binary is quite big at 108 MB's.

# Edit

The theory of skipping reactions affecting comments is wrong, because this migration included reactions and they're clearly present, but comments are **not** there:
http://forgej.ydns.eu/friendica/friendica2/issues/13391
...but I guess the small upside is that migrating reactions works when the page number is lowered, at least when comments are skipped. 

AFAIK there's no explicit option (besides code changes) to **not** migrate comments. They *should* always be migrated. Not sure what's going wrong.

**Comment by marcusxss** [comment]
Okay, I just did two imports from the same version/build - one friendica's repo and another a tiny test repo.

The tiny repo imported all comments.

For Friendica, SOME comments were migrated - specifically beginning on page 39 and then later. But earlier, they're not there:
http://forgej.ydns.eu/friendica/friendica2/issues?page=39&q=&type=all&sort=relevance&state=open&labels=&milestone=0&project=0&assignee=0&poster=0&archived=false

Maybe something breaks in the next comment it tries to migrate, or it's just only able to import X number of comments, or from X number of issues?
In two imports it has stopped in that exact spot.

Will investigate later today.

Edit: The comment migration stopped right at Tio's comment here (not inclusive): https://github.com/friendica/friendica/issues/10049#issuecomment-801191953

**Comment by MrPetovan** [comment]
Thank you for your continued work, does the comment import stops from the most recent or from the most ancient? If it's the former, it looks like it imports 5 years of comments, full stop (look at Tio's comment date). I know Friendica has a longer history, but maybe importing only 5 years of GitHub comments could be fine?

**Comment by marcusxss** [comment]
I'm not sure in which order it's done, but the ones that are migrated are from the most ancient and then forward, so all the most recent 40 pages or so of issues are missing comments. Initially I thought that maybe they were migrated from issue to issue and then all comments for each, since it's a sequence of issues that all have comments and then it stops, but then I found out some of the comments on the slightly older issues were missing, too. So maybe they are migrated chronologically, regardless of what issue they belong to.

I created a new issue about this problem:
https://codeberg.org/forgejo/forgejo/issues/11773

I also found out that once "maxPerPage" is lowered it's fine to migrate:
- Issues incl. reactions
- Releases 
- Labels
...so all of those were migrated too. Haven't tried Pull requests yet.

Meanwhile in 14.0.3 the Forgejo devs have solved the GitLab migration issue, so now I did a full migration from my GitLab import, without making any changes to the source, and it worked. So there's a full migration, around a week old, of issues, releases,  reactions and comments (from the look of it) here:
http://forgej.ydns.eu/friendica/friendica-gitlab/issues
...but since it's going through GitLab, the usernames are totally messed up.

**Comment by MrPetovan** [comment]
`import_user_test8474611_1fqG`, huh?

It's definitely encouraging, but we are back waiting on upstream fixes to complete our migration, which doesn't feel great.

**Comment by marcusxss** [comment]
I agree.

I don't have much of an update, but I'm trying to keep track of the Forgejo PR's getting in, and updating to the latest version and trying that. Forgejo 15 was released and I've tried migrating based on that, but there are still the same issues. I don't think v15 had anything migration related that hadn't been backported to V14, so from a migration standpoint I don't see a need to update git.friendi.ca yet.

The issues I've created hasn't had much response yet. I keep somewhat active in their Matrix channel to bring it up.

Saw a PR yesterday that fixes the broken "retry migration" option and that's a good UX improvement for migrations which I'm happy about, but it won't solve this issue.

Haven't looked further into debugging/fixing it myself yet, but I would still like to.
